File: 1201-1300/1216-Valid Palindrome III/1216-Valid Palindrome III.py
class Solution:
    def isValidPalindrome(self, s: str, k: int) -> bool:
        dp = [[0] * len(s) for _ in range(len(s))]
        for j in range(len(dp)):
            dp[j][j] = 1
            for i in range(j - 1, -1, -1):
                if s[i] == s[j]:
                    dp[i][j] = 2 + dp[i + 1][j - 1]
                else:
                    dp[i][j] = max(dp[i][j - 1], dp[i + 1][j])
        return dp[0][-1] >= len(s) - k

File: 1201-1300/1216-Valid Palindrome III/1216-Valid Palindrome III.cpp
class Solution {
public:
    bool isValidPalindrome(string s, int k) {
        vector<vector<int>> dp(s.size(), vector<int>(s.size()));
        for (int j = 0; j < s.size(); ++j) {
            dp[j][j] = 1;
            for (int i = j - 1; i >= 0; --i) {
                if (s[i] == s[j]) {
                    dp[i][j] = 2 + dp[i + 1][j - 1];
                }
                else {
                    dp[i][j] = max(dp[i][j - 1], dp[i + 1][j]);
                }
            }
        }
        
        return dp[0].back() >= s.size() - k;
    }
};
